### ISSUE TITLE
Possible deadlock with bidirectional streaming

### ISSUE THREAD

**Issue by exelsior** [grpc/grpc-dotnet#2576]
Hello.
Got interesting problem with bidirectional streaming through grpc connection in .Net.
I have three servers, like client <-> client-server <-> server.
client write in request stream requests, client-server process it (sort of), and server makes another requests to third party services (which is kinda slow). Problem occures when i write to request stream on client several dozens big requests (about 35-37 kb), server (through client-server) receives it, and starts process it. I think i should say that producer (client) is much faster than consumer (server).
Server relies on `await foreach IAsyncEnumerable`, so it reads first message, process it, write response to response stream and then read next request. But after processing first request, he can not write response to response stream, and consequently it can not read another request. I believe it's because of `GRPC_ARG_HTTP2_WRITE_BUFFER_SIZE ` parameter which is 65k by default. And if channel buffer is full of requests, there is literally no space for response, which is, basically, a deadlock. The most obvious solution - read all from request channel, store this request in some array and then just foreach it. But, unfortunately, this project has a use case, when server uses response data as a request for another service, and response is, basically, this data with some enrichments.
So with this case in mind, i can not use obvious solution, because in that case i would store all response in memory. And it can be hundreds of megabytes.
I tried to increase `InitialHttp2StreamWindowSize` in `SocketHttpHandler` which i use for `GrpcChannel` creation, to it maximum and it resolves the problem, but i realize it is a temporary solution. Setting `EnableMultipleHttp2Connections` in the same `SocketHttpHandler` doesn't help at all.
I created MRE for ensuring that this is not our code peculiarity
Is there some more permanent solution? Or i just do smth wrong?

I'm using Grpc.AspNetCore 2.63.0
OS: Macos Sonoma 14.1.1 
Device: Macbook pro m1
Dotnet: 8.0.300

MRE: https://drive.google.com/file/d/13JFrQp8KeSNZ6D8TQJjpKl5rNbaBYB7s/view?usp=drive_link

**Comment by JamesNK** [comment]
`GRPC_ARG_HTTP2_WRITE_BUFFER_SIZE` is used by Grpc.Core. It isn't a Grpc.AspNetCore or Grpc.Net.Client setting.

HTTP/2 flow control request and response use different buffers. A server that isn't reading request data fast enough shouldn't impact its ability to send response data.

https://medium.com/coderscorner/http-2-flow-control-77e54f7fd518

**Comment by exelsior** [comment]
Hello @JamesNK. 
Thank you for your answer. Different buffer - definitely good news. 
One less thing.
I tried to reduce message / increase InitialHttp2StreamWindowSize and it works.
But with the initial windowsize and relatively big message - it doesn't.
Do you have any ideas why server can not send his response?

**Comment by JamesNK** [comment]
The server might not be able to send the complete response because the buffer fills up, so it's waiting for the client to read it.

**Comment by exelsior** [comment]
I've ran out of ideas.
Seems like backpressure doesn't working.
https://github.com/exelsior/grpc-deadlock/tree/main

**Comment by exelsior** [comment]
Well. actually no. It's working. But because of full buffer, there is no update window and the whole system goes deadlock. Is it correct behavior?

**Comment by antontroshin** [comment]
We're experiencing the same behavior in Dapr Crypto API, behind the scenes, using bidirectional streaming to encrypt/decrypt data. https://github.com/dapr/dapr/issues/8244
It works on small payloads but gets stuck on ones around ~370-400Kb and more. This correlates with the first comment note, `several dozens big requests (about 35-37 kb)`.
During my attempt to fix the issue, I noticed that the "Receive" task was not being called as long as both "Send" and "Receive" were tied together either with `Task.WhenAll`, `.ContinueWith` or another variation of task chaining.
This led to a bad practice solution, of leaving the "Send" task hanging in the air without `await`ing for the task completion.
Would appreciate any idea/suggestion for a fix or confirmation that this indeed the deadlock happens and might be causing the issue.

**Comment by exelsior** [comment]
hey, @antontroshin 
i have no idea is actual deadlock this or not, but i managed to fix this issue with something like that.
`var taskWrite = Task.Run(async () =>
        {
            for (var i = 0; i < 50; i++)
            {
                await call.RequestStream.WriteAsync(new TestRequest()
                {
                    Message = message.ToString()
                }, cancellationToken);
            }
            
            await call.RequestStream.CompleteAsync();
        });

        var taskRead = Task.Run( async () =>
        {
            var q = 0;
            
            while (await call.ResponseStream.MoveNext())
            {
                Console.WriteLine(q);
                q++;
            }
        });
await Task.WhenAll(taskWrite, taskRead);`
Seems to me that you are right, and taskRead just didn't start.
It wait's till taskWrite (request to channel) overflow channel and there is no metaphorical room for answers.
Long story short this workaround is working perfectly for us.
Hope this will help you

**Comment by JamesNK** [comment]
I'm going to close this. I believe this is an application level issue, not a problem with the library or server infrastructure.
